### PR TITLE
torus-biconomy integration

### PR DESCRIPTION
**Pull request by divyaN73**
Total Biconomy API's integrated:
1. /wallet-user/getNonce : Get nonce for userLogin from Biconomy DB
2. /meta-tx/systemInfo : to get the Parameters & type to sign the message 
3. /wallet-login : to create a new contract wallet for a new user or return existing contract wallet address.
4. /getContractNonce : calculate/fetch current nonce value from SCW
5. /relay : relay transaction using biconomy relayer.

### CODE REVIEW

File: app/src/controllers/SmartContractWalletController.js [hunk 1]
```diff
@@ -44,20 +46,80 @@ export default class SmartContractWalletController {
...
+      })
+      if (response && response.flag == 200) {
+        return response.nonce
+      }
+      return
+    } catch (error) {
+      if (error.response.status == 404) {
+        return 0
+      }
+      return
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: This should also be return 0

File: app/src/controllers/SmartContractWalletController.js [hunk 2]
```diff
@@ -44,20 +46,80 @@ export default class SmartContractWalletController {
...
+    }
+  }
+
+  async messageToLogin(signer, nonce) {
+    let systemInfo = await get(`${config.biconomy}/api/v2/meta-tx/systemInfo?networkId=${KOVAN_CODE}`)
+
+    let message = {
+      userAddress: signer.toLowerCase(),
+      providerId: 100,
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: Move this to config.

File: app/src/controllers/SmartContractWalletController.js [hunk 3]
```diff
@@ -44,20 +46,80 @@ export default class SmartContractWalletController {
...
+
   /**
    * Request relayer to create a smart contract wallet
    */
   async createSmartContractWallet() {
     // Make an API call to relayer to create a smart contract wallet
     // Update the Vue state to processing
     try {
-      const obj = {
+      let obj = {
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: Keep it as const only

File: app/src/controllers/SmartContractWalletController.js [hunk 4]
```diff
@@ -44,20 +46,80 @@ export default class SmartContractWalletController {
...
+        nonce = 0
+      }
+      let _message = await this.messageToLogin(obj.owner, nonce)
+
+      let _privateKey = await this.getWallet(this.getSelectedEOA())
+      let _signature = sigUtil.signTypedMessage(new Buffer(_privateKey, 'hex'), { data: _message }, 'V3')
+      let data = {
+        signature: _signature,
+        from: obj.owner,
+        providerId: 100
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: use providerId value from config file.

File: app/src/config.js [hunk 5]
```diff
@@ -34,5 +34,16 @@ export default {
   redirect_uri: redirect_uri,
   api: 'https://backend.relayer.dev.tor.us',
   // api: 'https://api.tor.us',
-  relayer: 'https://relayer.relayer.dev.tor.us'
+  relayer: 'https://relayer.relayer.dev.tor.us',
+  biconomy: 'https://localhost:4000',
```
Comment (tomarsachin2271) [inline app/src/config.js]: keep the name as biconomyBaseURL

File: app/src/controllers/SmartContractWalletController.js [hunk 6]
```diff
@@ -69,6 +131,92 @@ export default class SmartContractWalletController {
...
+      response = await get(`${nonceURL}?signer=${signer}`, {
+        headers: {
+          'x-wallet-key': config.biconomyKey[KOVAN]
+        }
+      })
+      console.log(response)
+      return response.nonce
+    } catch (e) {
+      log.error(e)
+      return []
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: Why an empty array is returning. If there is an error, it should just return;

File: app/src/controllers/SmartContractWalletController.js [hunk 7]
```diff
@@ -69,6 +131,92 @@ export default class SmartContractWalletController {
...
+        headers: {
+          'x-wallet-key': config.biconomyKey[KOVAN],
+          'Content-Type': 'application/json; charset=utf-8'
+        }
+      })
+      console.log(response)
+      return response
+    } catch (e) {
+      log.error(e)
+      return []
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: why empty array is getting returned?

File: app/src/controllers/SmartContractWalletController.js [hunk 8]
```diff
@@ -69,6 +131,92 @@ export default class SmartContractWalletController {
...
+      baseGas: 0,
+      metaInfo: metaInfo,
+      relayerPayment: relayerPayment
+    }
+
+    let domainData = {
+      name: config.eip712DomainName,
+      version: config.eip712SigVersion,
+      verifyingContract: config.eip712VerifyingContract,
+      chainId: 42
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: Write a TODO comment to see where network id needs to be changed to get it dynamically based on current network selected.

File: app/src/controllers/SmartContractWalletController.js [hunk 9]
```diff
@@ -69,6 +131,92 @@ export default class SmartContractWalletController {
...
+      expiry: 0,
+      baseGas: 0,
+      metaInfo: metaInfo,
+      relayerPayment: relayerPayment
+    }
+
+    let domainData = {
+      name: config.eip712DomainName,
+      version: config.eip712SigVersion,
+      verifyingContract: config.eip712VerifyingContract,
```
Comment (tomarsachin2271) [inline app/src/controllers/SmartContractWalletController.js]: Set this value from systemInfo.relayHubAddress to get updated value as per the network selected.

File: app/src/controllers/TorusKeyring.js [hunk 10]
```diff
@@ -1,7 +1,8 @@
 import { EventEmitter } from 'events'
 import Wallet from 'ethereumjs-wallet'
 import * as ethUtil from 'ethereumjs-util'
-import sigUtil from 'eth-sig-util'
+// import sigUtil from 'eth-sig-util'
+const sigUtil = require('eth-sig-util')
```
Comment (tomarsachin2271) [inline app/src/controllers/TorusKeyring.js]: Make it to run with import style... check the eth-sig-util code to see how to import in in latest version

Comment (divyaN73) [reply app/src/controllers/TorusKeyring.js]: Tried, not working

**Comment by chaitanyapotti** [comment]
Closing due to inactivity

**Pull request closed without merging**